metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -155,12 +197,19 @@ class CheckpointingService[F[_]: Sync: Applicative, N](
 }
 
 object CheckpointingService {
-  type LedgerTree = Map[Block.Hash, (Ledger, Block.Header)]
+  type LedgerTree =
+    Map[Block.Hash, (Ledger, Block.Header, Option[CheckpointWithProof])]
```

Comment:
Alternatively we could just save the last proof along with the block hash in a separate Ref 🤔

Reply:
See this [comment](https://github.com/input-output-hk/metronome/pull/54/files/1a3f15eb976f58996f5249b52991d8585ac81380#r661533556), I reckon you would also need the headers along the way.

Reply:
Forget it. We need to maintain a tree and we don't really know which one is the _last one_ until we see a commit QC.

Reply:
The executed blocks are always the last ones, they are not going to be un-executed. Only validation is tentative. Execution is triggered by a commit of a descendant-or-self.

Reply:
Ok, you're right. Duh, I can move the accumulation outside of updateLedger. I will do that